RecoLocalMuon/GEMSegment/plugins/ME0SegAlgoRU.cc

```diff
@@ -0,0 +1,464 @@
+    	// to fit the segment
+    	float averageTime=0.;
+    	for(const auto* h : currentProtoSegment){
+    	  averageTime += h->rh->tof();
+    	}
+    	averageTime /= float(currentProtoSegment.size());
+    	float timeUncrt=0.;
+    	for(const auto* h : currentProtoSegment){
+    	  timeUncrt += pow( h->rh->tof()-averageTime,2);
+    	}
+    	timeUncrt = std::sqrt(timeUncrt/float(currentProtoSegment.size()-1));
+
+    	std::sort(currentProtoSegment.begin(),currentProtoSegment.end(),
+    			[](const HitAndPosition* a, const HitAndPosition *b) {return a->layer < b->layer;}
+    	);
+
+    	std::vector<const ME0RecHit*> bareRHs; bareRHs.reserve(currentProtoSegment.size());
+    	for(const auto* rh : currentProtoSegment) bareRHs.push_back(rh->rh);
+    	ME0Segment temp(bareRHs, current_fit->intercept(),
+    			current_fit->localdir(), current_fit->covarianceMatrix(), current_fit->chi2(), averageTime, timeUncrt);
+    	segments.push_back(temp);
+
+
+    }
+}
+
+void ME0SegAlgoRU::tryAddingHitsToSegment(const float maxTOF,  const float maxETA, const float maxPhi,const float maxChi2, std::unique_ptr<MuonSegFit>& current_fit, HitAndPositionPtrContainer& proto_segment,
+					  const BoolContainer& used, HitAndPositionContainer::const_iterator i1, HitAndPositionContainer::const_iterator i2) const {
+  // Iterate over the layers with hits in the chamber
+  // Skip the layers containing the segment endpoints
+  // Test each hit on the other layers to see if it is near the segment
+  // If it is, see whether there is already a hit on the segment from the same layer
+  //    - if so, and there are more than 2 hits on the segment, copy the segment,
+  //      replace the old hit with the new hit. If the new segment chi2 is better
+  //      then replace the original segment with the new one (by swap)
+  //    - if not, copy the segment, add the hit. If the new chi2/dof is still satisfactory
+  //      then replace the original segment with the new one (by swap)
+
+  //Hits are ordered by layer, "i1" is the inner hit and i2 is the outer hit
+  //so possible hits to add must be between these two iterators
+  for(auto iH = i1 + 1; iH != i2; ++iH ){
+	  if(iH->layer == i1->layer) continue;
+	  if(iH->layer == i2->layer) break;
+	  if(used[iH->idx]) continue;
+	  if(!areHitsConsistentInTime(maxTOF,proto_segment,*iH)) continue;
+	  if (!isHitNearSegment(maxETA,maxPhi,current_fit,proto_segment,*iH)) continue;
+	  if(hasHitOnLayer(proto_segment,iH->layer))
+		  compareProtoSegment(current_fit,proto_segment,*iH);
+	  else
+		  increaseProtoSegment(maxChi2, current_fit,proto_segment,*iH);
+  }
+}
+
+bool ME0SegAlgoRU::areHitsCloseInEta(const float maxETA, const bool beamConst, const GlobalPoint& h1, const GlobalPoint& h2) const {
+  // check that hits from different layer gets eta partition number +1 max..
+  float diff = 0;
+//  //actually no difference in our case...may want to consider something later
+//  if(beamConst){

```

Comment:
No, I removed it. I kept the beam constraint parameter, even though it is not used, since it will be used in the future when we optimize for displaced muons.